DEPR: Deprecate literal json string input to read_json

rmhowe425 opened the pull request:
- [X] closes #52271
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

rmhowe425 committed:
Adding logic to throw a deprecation warning when a literal json string is passed to read_json

rmhowe425 committed:
Adding logic to throw a deprecation warning when a literal json string is passed to read_json

rmhowe425 committed:
Updating documentation and adding PR num to unit test

rmhowe425 committed:
Adding a deprecation warning to the user guide

rmhowe425 committed:
Updating unit tests to check for FutureWarning

rmhowe425 committed:
Fixing unit tests

rmhowe425 committed:
Fixing unit tests

rmhowe425 committed:
Fixing unit tests

rmhowe425 committed:
Fixing unit tests

rmhowe425 commented:
Both failing tests appear to be failing because the tests pass json literal strings to `read_json`.

rmhowe425 commented:
@mroeschke Would you be able to take a look at this PR and let me know if anything needs updating?

mroeschke commented on doc/source/user_guide/io.rst:
```diff
@@ -2072,6 +2072,10 @@ is ``None``. To explicitly force ``Series`` parsing, pass ``typ=series``
 
+.. warning::
+
+    Passing json literal strings will be deprecated in a future release of pandas.
```
Can you move this to the `read_json` docstring instead?

rmhowe425 replied:
Should be fixed in latest push.

mroeschke reviewed (changes requested):
Could you change the existing tests by wrapping them in `StringIO`?

rmhowe425 committed:
Fixing documentation errors in PR feedback

rmhowe425 commented on doc/source/whatsnew/v2.1.0.rst:
```diff
@@ -268,9 +268,11 @@ Deprecations
 - Deprecated allowing arbitrary ``fill_value`` in :class:`SparseDtype`, in a future version the ``fill_value`` will need to be compatible with the ``dtype.subtype``, either a scalar that can be held by that subtype or ``NaN`` for integer or bool subtypes (:issue:`23124`)
 - Deprecated behavior of :func:`assert_series_equal` and :func:`assert_frame_equal` considering NA-like values (e.g. ``NaN`` vs ``None`` as equivalent) (:issue:`52081`)
 - Deprecated constructing :class:`SparseArray` from scalar data, pass a sequence instead (:issue:`53039`)
+- Deprecated literal json input to :func:`read_json`. Moving forward the method only accepts file-like objects (:issue:`53409`)
```
Should be fixed in latest push.

rmhowe425 commented on doc/source/whatsnew/v2.1.0.rst:
```diff
@@ -268,9 +268,11 @@ Deprecations
 - Deprecated positional indexing on :class:`Series` with :meth:`Series.__getitem__` and :meth:`Series.__setitem__`, in a future version ``ser[item]`` will *always* interpret ``item`` as a label, not a position (:issue:`50617`)
 -
 
+
```
Should be fixed in latest push.

mroeschke commented on doc/source/user_guide/io.rst:
```diff
@@ -2072,6 +2072,7 @@ is ``None``. To explicitly force ``Series`` parsing, pass ``typ=series``
 * ``engine``: Either ``"ujson"``, the built-in JSON parser, or ``"pyarrow"`` which dispatches to pyarrow's ``pyarrow.json.read_json``.
   The ``"pyarrow"`` is only available when ``lines=True``
 
+
```
```suggestion
```

rmhowe425 replied:
Should be fixed in latest push

mroeschke commented on pandas/io/json/_json.py:
```diff
@@ -535,6 +537,11 @@ def read_json(
         or ``StringIO``.
+
+        .. deprecated:: 2.1.0
+            Passing json literal strings will be deprecated in a future release of
```
```suggestion
            Passing json literal strings is deprecated
```

mroeschke commented on pandas/io/json/_json.py:
```diff
@@ -535,6 +537,11 @@ def read_json(
+
+        .. deprecated:: 2.1.0
+            Passing json literal strings will be deprecated in a future release of
+            pandas.
```
```suggestion
```

rmhowe425 replied:
Should be fixed in latest push

rmhowe425 committed:
Fixing documentation errors in PR feedback

rmhowe425 committed:
Updating unit tests to use StringIO rather than catch FutureWarning

rmhowe425 committed:
Finishing updating unit tests to use StringIO rather than catch FutureWarning

rmhowe425 commented:
@mroeschke This PR should be ready for review!

mroeschke commented on pandas/tests/io/json/test_compression.py:
```diff
@@ -23,7 +26,8 @@ def test_compression_roundtrip(compression):
         with tm.decompress_file(path, compression) as fh:
             result = fh.read().decode("utf8")
-        tm.assert_frame_equal(df, pd.read_json(result))
+            data = StringIO(result)
```
Can you keep the original indentation here?

mroeschke commented on pandas/tests/io/json/test_deprecated_kwargs.py:
```diff
@@ -10,9 +11,10 @@
 
 def test_good_kwargs():
     df = pd.DataFrame({"A": [2, 4, 6], "B": [3, 6, 9]}, index=[0, 1, 2])
-    with tm.assert_produces_warning(None):
```
Can you keep this here?

mroeschke commented on pandas/tests/io/json/test_readlines.py:
```diff
@@ -15,6 +15,20 @@
 from pandas.io.json._json import JsonReader
 
 
+def test_json_deprecation():
```
```suggestion
def test_literal_json_deprecation():
```

Also can you move this test to `test_pandas.py`?

rmhowe425 committed:
Fixing indendation errors in unit tests. Moved one unit test to another file.

rmhowe425 commented:
@mroeschke This PR should be ready for review!

rmhowe425 committed:
Updating unit test name

rmhowe425 committed:
Merge branch 'main' into dev/read_json

mroeschke commented on pandas/tests/io/json/test_pandas.py:
```diff
@@ -29,6 +29,20 @@
+    )
+
+    with tm.assert_produces_warning(FutureWarning, match=msg):
+        result = read_json('{"a": 1, "b": 2}\n{"b":2, "a" :1}\n', lines=True)
```
Could you test that this also raises a warning with `lines=False`

rmhowe425 replied:
@mroeschke I added a check in the JsonReader class in `_json.py` to check for the presence of "\n" when lines == False. 

The check seems to be working correctly. However, execution continues after a warning is fired and the code ends up triggering a Value Error: `ValueError: Trailing data`. Seems like the ValueError is overriding the warning and the warning is never caught.

Is there any way to catch the Warning in this scenario?

rmhowe425 replied:
@mroeschke So just to make sure that I understand, are you asking that add additional tests, or make sure that execution stops once a warning is fired? Or both?

rmhowe425 committed:
Adding additional checks to unit tests

rmhowe425 committed:
Merge branch 'dev/read_json' of github.com:rmhowe425/pandas into dev/read_json

rmhowe425 committed:
Fixing unit tests

rmhowe425 committed:
Merge branch 'main' into dev/read_json

rmhowe425 committed:
Fixing unit tests

rmhowe425 committed:
Merge branch 'main' into dev/read_json

rmhowe425 commented:
@mroeschke PR is ready for review. 

I believe I have addressed your concerns. From what I can tell, the test failures in Numpy Dev are not related to my PR

mroeschke commented on doc/source/whatsnew/v2.1.0.rst:
```diff
@@ -277,6 +277,7 @@ Deprecations
 - Deprecated allowing arbitrary ``fill_value`` in :class:`SparseDtype`, in a future version the ``fill_value`` will need to be compatible with the ``dtype.subtype``, either a scalar that can be held by that subtype or ``NaN`` for integer or bool subtypes (:issue:`23124`)
 - Deprecated behavior of :func:`assert_series_equal` and :func:`assert_frame_equal` considering NA-like values (e.g. ``NaN`` vs ``None`` as equivalent) (:issue:`52081`)
 - Deprecated constructing :class:`SparseArray` from scalar data, pass a sequence instead (:issue:`53039`)
+- Deprecated literal json input to :func:`read_json` (:issue:`53409`)
```
```suggestion
- Deprecated literal json input to :func:`read_json`. Wrap literal json string input in `io.StringIO` instead. (:issue:`53409`)
```

mroeschke commented:
Could you merge in the main branch too?

rmhowe425 committed:
Updating whatsnew documentation per reviewer recommendations.

rmhowe425 committed:
Merge branch 'main' into dev/read_json

rmhowe425 committed:
Merge branch 'main' into dev/read_json

mroeschke commented:
Looks like some of the code checks are still failing in the CI

rmhowe425 committed:
Fixing failing code tests

rmhowe425 committed:
Merge branch 'dev/read_json' of github.com:rmhowe425/pandas into dev/read_json

rmhowe425 committed:
Fixing failing code tests

rmhowe425 committed:
Adding import to doc string example

rmhowe425 committed:
Fixing documentation formatting error

rmhowe425 committed:
Fixing documentation formatting error

rmhowe425 commented:
@mroeschke Code Checks / Docstring validation passing now. Docbuild and upload test failure doesn't appear to be associated with PR

rmhowe425 committed:
Merge branch 'main' into dev/read_json

mroeschke commented:
You may want to check usage of `read_json` in the the `doc/source` folder for the deprecated behavor. PRs should still pass the doc build

rmhowe425 committed:
Fixing documentation error after fixing merge conflict

rmhowe425 committed:
Fixing formatting errors in whatsnew file

rmhowe425 committed:
Updating formatting errors in documentation

rmhowe425 committed:
Updating formatting errors in documentation

rmhowe425 commented:
@mroeschke Pinging on green!

rmhowe425 commented:
@mroeschke Should be good to go now!

mroeschke commented:
Thanks @rmhowe425
